### PR TITLE
PP-3679 Refactor to use PaymentSearchService in PaymentsResource

### PR DESCRIPTION
**Pull request by SandorArpa**
- move search logic from resource to service

### CODE REVIEW

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                    addLink(halRepresentation, "next_page", transformIntoPublicUri(baseUrl, searchResponse.getLinks().getNextPage()));
+                
+                    return halRepresentation.build();
+                } catch (IOException | ProcessingException | URISyntaxException ex) {
+                    throw new SearchChargesException(ex);
+                }
+            }
+            throw new SearchChargesException(connectorResponse);
+        }
+        throw new NotImplementedException("DD serch is not yet implemented");
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: spelling error `search`

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                    Pair.of(EMAIL_KEY, email),
+                    Pair.of(STATE_KEY, state),
+                    Pair.of(CARD_BRAND_KEY, cardBrand),
+                    Pair.of(FROM_DATE_KEY, fromDate),
+                    Pair.of(TO_DATE_KEY, toDate),
+                    Pair.of(TRANSACTION_TYPE_KEY, TRANSACTION_TYPE_KEY_VALUE),
+                    Pair.of(PAGE, pageNumber),
+                    Pair.of(DISPLAY_SIZE, displaySize)
+            );
+            javax.ws.rs.core.Response connectorResponse = client
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: qualified name can be replaced with import

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                    .request()
+                    .header(HttpHeaders.ACCEPT, APPLICATION_JSON)
+                    .get();
+
+            logger.info("response from connector form charge search: " + connectorResponse);
+
+            if (connectorResponse.getStatus() == SC_OK) {
+                try {
+                    JsonNode responseJson = connectorResponse.readEntity(JsonNode.class);
+                    logger.debug("json response from connector from charge search: " + responseJson);
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: I would remove this line as it was already logged above

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                                PaymentUriGenerator paymentUriGenerator,
+                                ObjectMapper objectMapper) {
+        this.client = client;
+        this.connectorUriGenerator = connectorUriGenerator;
+        this.paymentUriGenerator = paymentUriGenerator;
+        this.objectMapper = objectMapper;
+        this.baseUrl = configuration.getBaseUrl();
+    }
+    
+    public HalRepresentation doSearch(Account account, String reference, String email, String state, String cardBrand,
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: I would split this massive method into smaller ones. Too many things happening inside on method

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            throw new SearchChargesException(connectorResponse);
+        }
+        throw new NotImplementedException("DD serch is not yet implemented");
+    }
+
+    private String getConnectorUrl(Account account, List<Pair<String, String>> queryParams) {
+        String urlPath = connectorUriGenerator.chargesURI(account);
+        UriBuilder builder = UriBuilder
+                .fromPath(urlPath);
+        queryParams.stream().forEach(pair -> {
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: no need to `stream`, as `forEach` can be used directly on List

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        return account.getPaymentType().equals(DIRECT_DEBIT);
+    }
+
+    private void addLink(HalRepresentation.HalRepresentationBuilder halRepresentationBuilder, String name, URI uri) {
+        if (uri != null) {
+            halRepresentationBuilder.addLink(name, uri); 
+        }
+    }
+
+    private URI transformIntoPublicUri(String baseUrl, uk.gov.pay.api.model.links.Link link) throws URISyntaxException {
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: qualified name can be replaced with import

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        if (link == null)
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: Is there any case when this value can be null? Also, what happens if we return null?

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java (hunk 1 above)
Comment (georgeracu) [reply src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: Also should expand into Direct Debit as few people know that DD refers to Direct Debit

File: src/test/java/uk/gov/pay/api/service/PaymentSearchServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    @Before
+    public void setUp() {
+        connectorUriGenerator = new ConnectorUriGenerator(configuration);
+        paymentUriGenerator = new PaymentUriGenerator();
+        objectMapper = new ObjectMapper();
+        paymentSearchService = new PaymentSearchService(client, configuration, connectorUriGenerator, paymentUriGenerator, objectMapper);
+    }
+    
+    @Test
+    public void doSearchShouldThrowBadRequestException_whenAccountIsNotDD_andAgreementIsASearcParam() {
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/api/service/PaymentSearchServiceTest.java]: spelling error `Searc`

File: src/test/java/uk/gov/pay/api/service/PaymentSearchServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import uk.gov.pay.api.auth.Account;
+import uk.gov.pay.api.model.TokenPaymentType;
+
+import javax.ws.rs.client.Client;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PaymentSearchServiceTest {
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/api/service/PaymentSearchServiceTest.java]: No happy path test?

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java (hunk 1 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: It's for one iteration but will change it

File: src/test/java/uk/gov/pay/api/service/PaymentSearchServiceTest.java (hunk 9 above)
Comment (SandorArpa) [reply src/test/java/uk/gov/pay/api/service/PaymentSearchServiceTest.java]: All happy path tests are covered in `PaymentResourceSearchITest`

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java (hunk 7 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: I don't know, this code is mainly copy paste from `PaymentsResource`. I assume there is a reason why it is how it is.

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java (hunk 4 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: Next iteration perhaps, now I'm just moving everything into a service

**Pull request merged**